src/pip/_internal/resolution/resolvelib/factory.py

```diff
@@ -264,9 +264,36 @@ def find_candidates(
             if ireq is not None:
                 ireqs.append(ireq)
 
+        for link in constraint.links:
+            if not ireqs:
+                # If we hit this condition, then we cannot construct a candidate.
+                # However, if we hit this condition, then none of the requirements
+                # provided an ireq, so they must have provided an explicit candidate.
+                # In that case, either the candidate matches, in which case this loop
+                # doesn't need to do anything, or it doesn't, in which case there's
+                # nothing this loop can do to recover.
+                break
+            ireq = ireqs[0]
+            if link.is_wheel:
+                wheel = Wheel(link.filename)
+                # Check whether the provided wheel is compatible with the target
+                # platform.
+                if not wheel.supported(self._finder.target_python.get_tags()):
+                    return ()
```

Comment:
Why is this check needed? Does the explicit candidate code path have this? (I don’t recall it does.)

Reply:
The equivalent code lives in `make_requirement_from_install_req`, but that code raises an exception because it's committed to trying to install the wheel. I figured the right behavior for a constraint would be to force backtracking to see if there's a set of dependencies that doesn't install the constrained package.

(I mean, if it comes to this, something has clearly gone *wrong*, but this is how I interpreted "If this package is going to be installed, it has to be this version." -> "If this version can't be installed, don't install the package.")

Reply:
Thanks, this makes sense for me now.